tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -37,6 +37,92 @@ struct EntryCluster {
    Long64_t end;
 };
 
+// note that this routine assumes global entry numbers
+static bool ClustersAreSortedAndContiguous(const std::vector<std::vector<EntryCluster>> &cls)
+{
+   Long64_t last_end = 0ll;
+   for (const auto &fcl : cls) {
+      for (const auto &c : fcl) {
+         if (last_end != c.start)
+            return false;
+         last_end = c.end;
+      }
+   }
+   return true;
+}
+
+/// Take a vector of vectors of EntryClusters (a vector per file), filter the entries according to entryList, and
+/// and return a new vector of vectors of EntryClusters where cluster start/end entry numbers have been converted to
+/// TEntryList-local entry numbers.
+///
+/// This routine assumes that entry numbers in the TEntryList (and, if present, in the sub-entrylists) are in
+/// ascending order, i.e., for n > m:
+///   elist.GetEntry(n) + tree_offset_for_entry_from_elist(n) > elist.GetEntry(m) + tree_offset_for_entry_from_elist(m)
+static std::vector<std::vector<EntryCluster>>
+ConvertToElistClusters(std::vector<std::vector<EntryCluster>> &&clusters, TEntryList &entryList,
+                       const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                       const std::vector<Long64_t> &entriesPerFile)
+{
+   R__ASSERT(entryList.GetN() > 0); // wasteful to call this function if it has nothing to do
+   R__ASSERT(ClustersAreSortedAndContiguous(clusters));
+
+   const bool listHasGlobalEntryNumbers = entryList.GetLists() == nullptr;
+   const auto nFiles = clusters.size();
+
+   std::unique_ptr<TChain> chain;
+   if (!listHasGlobalEntryNumbers) {
+      // we need to build `chain` to be able to convert local entry numbers to global entry numbers in `Next`, below
+      chain.reset(new TChain());
+      for (auto i = 0u; i < nFiles; ++i)
+         chain->Add((fileNames[i] + "/" + treeNames[i]).c_str(), entriesPerFile[i]);
+   }
+
+   // Advance elEntry, then return global entry number corresponding to the elEntry-th entry in elist
+   // Return -1 if we are out of TEntryList entries.
+   auto Next = [listHasGlobalEntryNumbers, &chain](Long64_t &elEntry, TEntryList &elist) {
+      ++elEntry;
+      if (listHasGlobalEntryNumbers) {
+         return elist.Next();
+      } else {
+         int treenum = -1;
+         Long64_t globalEntry = elist.GetEntryAndTree(elEntry, treenum);
+         if (globalEntry != -1ll)
+            globalEntry += chain->GetTreeOffset()[treenum];
+         return globalEntry;
+      }
+   };
```

Comment:
```suggestion
   auto Next = [listHasGlobalEntryNumbers, &chain](Long64_t &elEntry, TEntryList &elist) {
      ++elEntry;
      return elist.Next();
  }
  if (!listHasGlobalEntryNumbers) 
      Next =  [listHasGlobalEntryNumbers, &chain](Long64_t &elEntry, TEntryList &elist) {
         ++elEntry;
         int treenum = -1;
         Long64_t globalEntry = elist.GetEntryAndTree(elEntry, treenum);
         if (globalEntry != -1ll)
            globalEntry += chain->GetTreeOffset()[treenum];
         return globalEntry;
     };
```
(assuming it compiles)

Reply:
every lambda has a different type, so that can't compile. I could make `Next` a `std::function`, but that requires bringing in the header, and it basically makes `Next()` a virtual call, that might erase the benefit of removing the branching..

Reply:
> every lambda has a different type, so that can't compile.

That's a pity ... (and I suppose ```auto Next = listHasGlobalEntryNumbers ? ... : ...;``` wont work either :(

Reply:
nope -- if the capture list was empty, the lambda type would have a conversion operator to function pointer that we could use to normalize the types, but that's not the case either

Reply:
According to https://godbolt.org/z/DWGxmM the operator ```? :``` as a chance of working.

Reply:
because the capture list is empty, see my last comment

Reply:
humm then the following works.
```suggestion
   auto Next = listHasGlobalEntryNumbers ?
    [](Long64_t &elEntry, TEntryList &elist, TChain *) {
         ++elEntry;
         return elist.Next();
     } 
    :
     [](Long64_t &elEntry, TEntryList &elist, TChain *chain) {
         ++elEntry;
         int treenum = -1;
         Long64_t globalEntry = elist.GetEntryAndTree(elEntry, treenum);
         if (globalEntry != -1ll)
            globalEntry += chain->GetTreeOffset()[treenum];
         return globalEntry;
     };
```
(whether it is better than the original is still debatable :) )

Reply:
how does it look now?